---
title: "HiPathia Bioconductor Package"
author: "Marta R. Hidalgo, Francisco Salavert, Alicia Amadoz, Çankut Cubuk, José Carbonell-Caballero, Joaquín Dopazo"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

HiPathia is a method for the computation of signal transduction along
signaling pathways. The method is based on an iterative algorithm which
is able to compute the signal intensity passing through the nodes of a
network by taking into account the level of expression of each gene and
the intensity of the signal arriving to it.


## Introduction

HiPathia package implements the Canonical Circuit Activity Analysis method
for the quantification of the signaling pathways activity presented in 
[Hidalgo et al.](https://www.ncbi.nlm.nih.gov/pubmed/28042959). It has been
conceived as a functional tool for R users which allows more control on the
analysis pipeline than the web implementation http://hipathia.babelomics.org.

This document will introduce you to the HiPathia package and how to use it to
analyze your data. 



## Previous considerations

HiPathia is a method for the computation of signal transduction along
signaling pathways. The method is independent on the pathways database, it only 
needs information about the topology of the graph and the genes included in each
node. 

However, due to computational cost, HiPathia needs to preprocess the graphs to 
be fully efficient. In the current implementation we have developed a module 
which has preprocessed 145 KEGG pathway KGML files, which are ready to be 
analyzed.

Further versions of the package will allow the user to preprocess their own 
graph pathways to be analyzed with HiPathia.


### Instalation 

In order to install the HiPathia package, type on your R console

```{r, fig.show='hold', message=FALSE, warning=FALSE}
## try http:// if https:// URLs are not supported 
source("https://bioconductor.org/biocLite.R") 
biocLite("hipathia")
```



### Example data

In order to illustrate the HiPathia package functionalities an example dataset
has been prepared. Data has been downloaded from The Cancer Genome Atlas data
repository, from the BRCA-US project. 20 tumor and 20 normal samples of RNA-Seq
data have been randomly selected and normalized. Specifically, data has been
corrected for batch effect using the COMBAT method, then corrected for RNA
composition bias applying TMM normalization and log-transformed.

```{r, fig.show='hold', message=FALSE, warning=FALSE}
library(hipathia)
data("brca_data")
```

The information about the type of these samples has been stored in a
different dataset, and can be loaded in the same way as before.

```{r, fig.show='hold', message=FALSE, warning=FALSE}
data("brca_design")
```


## Preprocessment

### Preprocess data

HiPathia accepts as input data a gene expression matrix. Expression may have
been measured with any available sequencing technique. However, HiPathia
assumes that data has been already normalized for correcting any possible
sequencing bias (which includes also batch effect correction). 

```{r, fig.show='hold',}
hhead(brca_data)
```


The gene expression matrix must include samples as columns and genes as rows,
as shown in the previous example. Rownames must be the Entrez IDs of the genes
in the rows. In order to transform other gene IDs to Entrez IDs, function
`translate.matrix` can be used. Accepted IDs to be transformed to Entrez IDs
include

**Human**
* Affy HG U133A probeset
* Affy HG U133B probeset
* Affy HG U133-PLUS_2 probeset
* Agilent SurePrint G3 GE 8x60k
* Agilent SurePrint G3 GE 8x60k v2
* Agilent Whole Genome 4x44k
* Agilent Whole Genome 4x44k v2
* CCDS
* Ensembl gene
* Ensembl transcript
* Entrez ID
* GenBank EMBL
* GenBank PID
* HGNC symbol
* RefSeq mRNA
* RefSeq mRNA PRED
* RefSeq ncRNA
* RefSeq ncRNA PRED

**Mouse**
* Affy Mouse 430 2
* Ensembl gene
* Gene name
* Mouse Gene 1.0

**Rat**
* Ensembl gene
* Gene name


The parameters needed by this function are the data matrix and the species
of the experiment.

```{r, fig.show='hold'}
trans_data <- translate.matrix(brca_data, "hsa")
```

Apart from the necessary bias corrections, expression data must be normalized
between 0 and 1 before applying the `hipathia` function. Function 
`normalize.data` is designed to this purpouse. 

```{r, fig.show='hold'}
boxplot(trans_data)
exp_data <- normalize.data(trans_data)
boxplot(exp_data)
```

Function `normalize.data` includes different parameters for normalization.
If option `by.quantiles` is `TRUE`, a previous normalization by quantiles
is performed.

```{r, fig.show='hold'}
exp_data <- normalize.data(trans_data, by.quantiles = TRUE)
boxplot(exp_data)
```

Other parameters of this function affect the way in which normalization
to the interval [0,1] is performed. Parameter `by.gene` indicates whether
to perform the transformation [0,1] to each row of the matrix. If the option
`by.gene` is set to TRUE, the normalization between 0 and 1 is done for each
row of the matrix, meaning that the expression of each gene will have a range
between 0 and 1. If it is set to FALSE, the normalization is done for the 
whole matrix, meaning that only the genes with the maximum value of the 
matrix will have a normalized value of 1. It is recommended to keep it set
to `FALSE`, as the default value. 
 
Parameter `percentil` indicates whether to use the percentil to compute the
normalized value between 0 and 1. If it is set to TRUE, the function takes as
a value for the position `(i,j)` of the matrix the percentil of sample `j` in
the ditribution of gene `i`. If it is seet to FALSE, the functions applies a
direct transformation from the original interval to [0,1]. It is recommended
to keep it set to `FALSE` except for heavy-tailed distributions of the genes. 

```{r, fig.show='hold'}
exp_data <- normalize.data(trans_data, percentil = TRUE)
boxplot(exp_data)
```

Parameter `truncation.percentil` gives the value of percentil `p` from which
all further values are truncated to percentil `p`. Symmetrically, values beyond
percentil `1-p` are also truncated to `1-p`.

```{r, fig.show='hold'}
exp_data <- normalize.data(trans_data, truncation.percentil = 0.95)
boxplot(exp_data)
```



### Load pathways

HiPathia package is currently implemented to use preprocessed KEGG pathways.
The pathways have been processed and stored in a pathways object. This object
includes all the information that the different functions in the package need.
In order to load this object, use function `load.pathways` and select the
species to be analyzed. Available species include human ("hsa"), mouse ("mmu")
and rat ("rno").

```{r, fig.show='hold'}
pathways <- load.pathways(species = "hsa")
```

Parameter `pathways.list` allows the user to specify the pathways to be loaded.
The different functions of the package will use all the pathways in the 
pathways object for its computations. In order to restrict the analysis to a
particular set of pathways, load only the required pathways to the pathway
object. By default, all pathways available for the specified species are
loaded.

```{r, fig.show='hold'}
pathways_only2 <- load.pathways(species = "hsa", pathways.list = c("hsa03320",
                                                                   "hsa04014"))
```

In order to know which pathways are included in each pathways object,
function `get.pathways.list` can be used.

```{r, fig.show='hold'}
length(get.pathways.list(pathways))
get.pathways.list(pathways)
```

```{r, fig.show='hold'}
length(get.pathways.list(pathways_only2))
get.pathways.list(pathways_only2)
```

## Pathway activation computation

HiPathia aims to compute the level of activation of each subpathway in a
pathway for each of the samples from the experiment. This is done by function
`hipathia`, which takes as inputs the matrix of gene expression, the pathways
object and some other parameters.

Parameter `decompose` indicates whether to use effector subpathways or to
decompose the pathways into decomposed subpathways. For further information
on this, see next section **Subpathways**. For further information on the
method used to compute the level of signal activity in the pathways, see
Section **Computing the signal** or refer to 
[Hidalgo et al.](https://www.ncbi.nlm.nih.gov/pubmed/28042959).

```{r, fig.show='hold'}
results <- hipathia(exp_data, pathways, decompose = FALSE, verbose=FALSE)
```

The resulting object includes the level of activity of the signal in each
subpathway as well as further information of the computation. In order to
extract the matrix of signal activity values from this object use function
`get.paths.matrix`.

```{r, fig.show='hold'}
path_vals <- get.paths.matrix(results)
hhead(path_vals)
```

Rownames of the matrix of pathway results are the IDs of the subpathways.
In order to transform this ID to a comprehensive name, use `get.path.names`
function. However, it is not recommended to change the row names of this
matrix.

```{r, fig.show='hold'}
path_names <- get.path.names(pathways, rownames(path_vals))
head(path_names)
```

Notice that the matrix of subpathway activity values will include a value of
activity for each sample and for each possible subpathway of the pathways in
the pathway object. Depending on whether parameter `decompose` is set to TRUE
or FALSE, and on the number of pathways included in the object of pathways
given as attribute, the number of analyzed subpathways may vary. Currently
HiPathia includes up to the following number of pathways, effector subpathways
and decomposed subpathways per species: 


```{r, echo=FALSE, results='asis'}
tab <- t(sapply(c("hsa", "mmu", "rno"), function(species){
   p <- load.pathways(species)
   effs <- sum(sapply(p$pathigraphs, function(pathi) length(
       pathi$effector.subgraphs)))
   decs <- sum(sapply(p$pathigraphs, function(pathi) length(pathi$subgraphs)))
   n <- length(p$pathigraphs)
   c(n, decs, effs)
}))
colnames(tab) <- c("Pathways", "Effector subpathways", "Decomposed subpathways")
knitr::kable(tab)
```


### Subpathways

Pathways are represented by directed graphs, which include different input
and output nodes. The signal arrives to an initial node and is transmited
along the pathway following the direction of the interactions up to an output
node. Thus, the signal may follow many different paths along the pathway.
HiPathia computes the intensity of this signal up to each output node of a
pathway separately.

Genes in the output nodes are also called *effector proteins*, since they
are the ones responsibles for performing the action the signal is seeking.
We define the *effector subpathway* ending in node *G* as the subgraph
including any node in a path leading to *G*. When applied to effector
subpathways, HiPathia returns the intensity of the signal arriving to the
effector protein *G*.

Effector subpathways may have many different input nodes. In order to analyze
in detail which of the possible paths leading to node *G* is responsible for
the observed change, effector subpathays can be decomposed into several
subpathways including only one input node. We define the 
*decomposed subpathway* from *H* to *G* as the subgraph including any node
in a path from *H* to *G*.

It is recommended to perform an initial HiPathia analysis with effector
subpathways, and use decomposed subpathways only for specific pathways in
which the user is highly interested.

### Computing the signal

In order for a protein to pass the signal, there are two important factors:
first, the protein must be present, and second, some other protein must
activate it. Therefore, HiPathia is a method to compute signal transduction
based on two steps. First, it quantifies the presence of a particular gene
as a normalized value between 0 and 1. Then, it computes the signal value
passing through a node taking into account the level of expression of each
gene inside the node and the intensity of the signal arriving to it. The
signal value of the pathway is the signal value through the last node of the
pathway. 

#### Node expression

Pathways are represented by graphs and composed by nodes and relations
among them. Some nodes may contain multiple genes representing different
isoforms of the protein or members of the same gene familiy, among others.
Since each gene has its own level of expression, the first step of the
method is to summarize this information into a score representing the
expression of the node as a whole. 

#### Signal transduction

The computation of the signal intensity across the pathway is performed by
means of an iterative algorithm beginning in the input nodes of the
subpathway. In order to initialize the pathway signal we assume an incoming
signal value of 1 in the input nodes of the subpathway. Then, for each node
$n$ of the network, the signal value $S_n$ is propagated along the nodes 
according to the following rule
\begin{equation}
\label{formula}
 S_n = v_n\cdot(1-\prod_{s_i \in A_n}(1-s_i))\cdot\prod_{s_j\in I_n}(1-s_j)
\end{equation}

\noindent where $A_n$ is the set of signals arriving to the current node from an
activation edge, $I_n$ is the set of signals arriving to the current node from 
an inhibition edge, and $v_n$ is the normalized value of expression of the 
current node. 


## Function activation computation

Each effector protein of a pathway is responsible for performing a particular
function. Thus, from the matrix of effector subpathways we can infer the
functions matrix, which gives an intensity value for each molecular function
and for each sample. 

Different effector subpathways of different pathways may end in the same
effector protein, and also different effector genes may have the same molecular
function. Therefore, for a particular function $f$, we summarize the values of
all the subpathways ending in an effector protein related to $f$ with a mean
value.

Different function activity matrices can be computed depending on the
functional annotation given to the effector nodes. Function `quantify.terms`
accepts any annotation defined by the user and it has also two default
annotations: Gene Ontology functions and Uniprot keywords. 

```{r, fig.show='hold'}
uniprot_vals <- quantify.terms(results, pathways, "uniprot")
go_vals <- quantify.terms(results, pathways, "GO")
```

The result of this function is a matrix with the level of activity of each
annotated function for each sample. Notice that functions annotated to genes
which are not included in any effector node will be not computed.


## Pathway/Function activation analysis

Once the matrix of desired features has been computed, any kind of analysis
may be performed on it, in the same way as if it were the matrix of gene
expressions. Specifically, comparison of the features across different groups
of samples is one of the keys. We can perform a comparison of two groups
applying the Wilcoxon test using function `do.wilcoxon`.

```{r, fig.show='hold'}
sample_group <- brca_design[colnames(path_vals),"group"]
comp <- do.wilcoxon(path_vals, sample_group, g1 = "Tumor", g2 = "Normal")
hhead(comp)
```

Function `get.pathways.summary` returns a summary by pathway of the results
from the Wilcoxon test, summaryzing the number of significant up- or
down-activated features.

```{r, fig.show='hold'}
pathways_summary <- get.pathways.summary(comp, pathways)
head(pathways_summary, n = 15)
```

Principal Components Analysis can be also performed by using function
`do.pca`. Notice that the number of rows must not exceed the number of
columns of the input matrix.

```{r, fig.show='hold'}
ranked_path_vals <- path_vals[order(comp$p.value, decreasing = FALSE),]
pca_model <- do.pca(ranked_path_vals[1:ncol(ranked_path_vals),])
```

PCA models can be visualized with a specific function called `pca.plot`.
See section **Visualization** for further information.


## Visualization

### Heatmap

Function `heatmap.plot` plots a heatmap with the values of the given matrix.
The experimental design can be provided to assign a class to each sample by
means of the parameter `sample.type`. Notice that the calsses must be in the
same order as the columns of the provided matrix. One can select whether to
cluster samples or variables with parameters `variable.clust` and
`sample.clust`. The colors of the different classes of samples can be selected
through parameter `sample.colors` with a vector of colors named after the
classes. The colors inside the heatmap can be also selected with parameter
`colors`. Personalized colors can be provided as a vector, or preselected
color schemes *classic* (default), *hipathia* or *redgreen* may be chosen. 

```{r, fig.show='hold'}
heatmap.plot(path_vals, sample.type = sample_group)
heatmap.plot(path_vals, sample.type = sample_group, colors="hipathia", 
             variable.clust = TRUE)
```


```{r, fig.show='hold'}
heatmap.plot(ranked_path_vals[1:15,], sample.type = sample_group, 
             variable.clust = TRUE)
heatmap.plot(ranked_path_vals[1:15,], sample.type = sample_group, 
             colors="hipathia", variable.clust = TRUE)
heatmap.plot(ranked_path_vals[1:15,], sample.type = sample_group, 
             colors="redgreen", variable.clust = TRUE)
```


### PCA

Function `pca.plot` plots two components of a PCA model computed with function
`do.pca` (see section **Pathway/Function activation analysis**). The
experimental design can be provided to assign a class to each sample by means
of the parameter `sample.type`. Notice that the classes must be in the same
order as the columns of the matrix provided to the PCA model. The colors of
the different classes of samples can be selected through parameter
`sample.colors` with a vector of colors named after the classes. If no such
parameter is provided, a predefined set of colors will be assigned. A main
title may be given to the plot through parameter `main`. The components to be
plotted can be selected through parameters `cp1` and `cp2` giving integer
number. If parameter `legend` is set to TRUE, the legend will be plotted.

```{r, fig.show='hold'}
pca.plot(pca_model, sample_group, legend = TRUE)
pca.plot(pca_model, sample.type = rep(1:5, 8), main = "Random types", 
         legend = TRUE)
```

Function `multiple.pca.plot` plots $n$ PCA components given by parameter
`comps` as an integer vector. As before, the experimental design can be
provided to assign a class to each sample by means of the parameter
`sample.type`. Notice that the classes must be in the same order as the
columns of the matrix provided to the PCA model. The colors of the different
classes of samples can be selected through parameter `sample.colors` with a
vector of colors named after the classes. If no such parameter is provided,
a predefined set of colors will be assigned. The cumulative explained variance
can be represented by setting `plot.variance` parameter to TRUE. If parameter
`legend` is set to TRUE, the legend will be plotted. A main title may be given
to the plot through parameter `main`. 

```{r, fig.show='hold'}
multiple.pca.plot(pca_model, sample_group, plot.variance = FALSE, 
                  legend = FALSE)
multiple.pca.plot(pca_model, sample_group, cex=3, plot.variance = TRUE)
multiple.pca.plot(pca_model, sample.type = rep(1:5, 8), 
                  main = "Random sample types")
```


### Pathway comparison

The results of a comparison are sometimes difficult to summarize. An easy
way to understand these results is to visualize them as an image. Function
`pathway.comparison.plot` creates an image of a pathway, with the same layout
from KEGG, including a color code representing the significant up- and
down-activated subpathways, and, if desired, the significant up- and
down-regulated nodes. 

```{r, fig.show='hold'}
pathway.comparison.plot(comp, metaginfo = pathways, pathway = "hsa03320")
pathway.comparison.plot(comp, metaginfo = pathways, pathway = "hsa04014")
```

In these plots, colored edges represent significant subpathways. Edges
belonging to subpathways which are significantly down-activated will be
depicted in blue and those belonging to subpathways which are significantly
up-activated will be depicted in red (as default). The *up* and *down* colors
may be changed by the user through the parameter `colors` by giving a vector
with three colors (representing down-activation, non-significance and
up-activation respectively) or a color scheme (either *classic* or *hipathia*).

In order to visualize the effect of the nodes expression differences in the
pathways, nodes can be colored by its differential expression. The color of
each node with respect to its differential expression must be previously
computed using function `node.color.per.de`. Note that
this fucntion computes differential expression on the nodes, not on the genes.
It uses function `eBayes` from package `limma`, see the `limma` package
vignette for further information. 

When computed, the resulting object must be provided to the
`pathway.comparison.plot` function as parameter `node.colors`.

```{r, fig.show='hold'}
colors_de <- node.color.per.de(results, pathways, sample_group, "Tumor", 
                               "Normal")
pathway.comparison.plot(comp, metaginfo = pathways, pathway = "hsa03320", 
                        node.colors = colors_de)
colors_de_hipathia <- node.color.per.de(results, pathways, sample_group, 
                                        "Tumor", "Normal", colors = "hipathia")
pathway.comparison.plot(comp, metaginfo = pathways, pathway = "hsa04014", 
                        node.colors = colors_de_hipathia, colors = "hipathia")
```


### Visualization through a local server

HiPathia results can be viewed on a web browser interactivelly. In order to
save the files for their visualization, use function `create.report`. The
parameters to be provided are the object of results, the Wilcoxon comparison,
the pathways object and the path to the folder in which the files must be
saved. Optionally, the colors of the nodes can be also provided using an
object returned by function `color.nodes.per.differential.expression` or a
similar data structure. 

```{r, fig.show='hold'}
create.report(comp, pathways, "save_noColors")
create.report(comp, pathways, "save_colors", node.colors = colors_de)
```

Due to cross-origin security restrictions 
([CORS](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing)), 
a web server is needed to serve the result files correctly.
The easiest way to run a web server to show the result files is with the
HiPathia function `visualize.report`. The user must specify the folder where
the report has been stored by function `create.report`. A web server developed
in R will be executed, serving the result files to the default URL
http://127.0.0.1:4000. Port 4000 may be changed through parameter `port`. 

```{r, fig.show='hold'}
visualize.report("save_colors")
```

```{r, fig.show='hold'}
visualize.report("save_noColors", port = 4001)
```

The function `visualize.report` uses the `servr` package, please refer to
[servr](https://github.com/yihui/servr) for further information.

The servers will be active until function `daemon_stop` from package `servr`
is executed. Information about how to stop each server individually is given
as an output of each `visualize.report` function. To stop all servers at a
time, use

```{r, fig.show='hold'}
servr::daemon_stop()
```


Alternatively, if you have already a web server installed in your computer,
just link or move the output folder to your web server http document root
and then open it on your web browser.


### Visualization through a local server with different groupings

Effector subpathway results are shown by default grouped by the pathway to which
they belong. However, if our interest is to see the comparison of all the 
subpathways arriving to a particular function, we can group the subpathways by
Uniprot or GO functions. Moreover, if we want to see the results of all the 
subpathways containing a particular gene, we can group the subpathways by
genes.

In order to do that, we must use the `group.by` parameter of functions 
`color.node.per.differential.expression` and `create.report`. 
Available `group.by` parameter values include: `uniprot`, to group subpathways 
arriving to the same Uniprot functions, `GO`, to group subpathways arriving to 
the same GO terms, and `genes`, to group subpathways containing each particular 
gene.

```{r, fig.show='hold'}
colors_de_uni <- node.color.per.de(results, pathways, sample_group, "Tumor", 
                                   "Normal", group.by = "uniprot")
create.report(comp, pathways, "save_colors_uniprot", 
              node.colors = colors_de_uni, group.by = "uniprot")
visualize.report("save_colors_uniprot", port = 4002)
```

As before, to stop the server and free the port, use the information about how 
to stop each server individually in the output of each `visualize.report` 
function or stop all servers at a time, using

```{r, fig.show='hold'}
servr::daemon_stop()
```


#### Interpreting HTML results

The interactive visualization of HiPathia results includes three panels and
a legend. The legend is on top of the page resuming the main information
depicted in the images. The left panel is the pathways panel, where the
currently selected pathway is shown. The layout of the pathway is similar
to the layout shown in KEGG. 

As before, edges belonging to significant down-activated pathways are depicted
in blue, those belonging to significant up-activated subpathways are depicted
in red, and those belonging to non-significant subpathways are depicted in
grey. Similarly, when nodes are colored by their differential expression,
down-regulated nodes are colored in blue, up-regulated nodes are colored in
red and non-significant nodes are colored in white. Different shades of the
colors indicate different levels of significance with respect to the p-value
of the differential expression. 

The selected pathway to be shown can be modified through the pathway list in
the top right panel. Arrows pointing up and down to the left of the names of
the pathways indicates that the pathways contain up- or down-activated
subpathways, respectively. When the arrows are colored in red or blue, it
means that there are significant up- or down-regulated subpathways,
respectively. The pathways list can be filtered through the *Filter...* box,
or ordered by means of the buttons in the top right part of the panel.

All computed subpathways of the currently selected pathway are listed in the
subpathways list in the bottom right panel. Arrows pointing up and down by the
names of the subpathways indicates that they are up- or down-activated,
respectively. When the arrows are colored in red or blue, it means that they
are significantly up- or down-regulated, respectively. When a subpathway is
selected from the list, only the arrows and nodes belonging to this subpathway
will be highlighted. Clicking again on this subpathway will deselect it.



